pkg/syncer/cnsoperator/controller/cnsunregistervolume/cnsunregistervolume_controller.go

```diff
@@ -177,7 +181,198 @@ func (r *ReconcileCnsUnregisterVolume) Reconcile(ctx context.Context,
+		Names: []string{
+			metadata,
+		},
+	}
+
+	queryResult, err := r.volumeManager.QueryVolumeAsync(ctx, queryFilter, &querySelection)
+	if err != nil {
+		msg := fmt.Sprintf("Unable to query volume %q . Error: %+v", instance.Spec.VolumeID, err)
+		log.Error(msg)
+		setInstanceError(ctx, r, instance, msg)
+		return reconcile.Result{RequeueAfter: timeout}, nil
+	}
+	if len(queryResult.Volumes) == 0 {
+		log.Infof("Volume: %q not found while querying CNS", instance.Spec.VolumeID)
+		return reconcile.Result{RequeueAfter: timeout}, nil
+	}
+
+	cnsVol := queryResult.Volumes[0]
+
+	var pvName, pvcName, namespace string
+	for _, entity := range cnsVol.Metadata.EntityMetadata {
+		entityType := entity.(*cnstypes.CnsKubernetesEntityMetadata).EntityType
+		if entityType == string(cnstypes.CnsKubernetesEntityTypePV) {
+			pvName = entity.(*cnstypes.CnsKubernetesEntityMetadata).EntityName
+		}
+
+		if entityType == string(cnstypes.CnsKubernetesEntityTypePVC) {
+			pvcName = entity.(*cnstypes.CnsKubernetesEntityMetadata).EntityName
+			namespace = entity.(*cnstypes.CnsKubernetesEntityMetadata).Namespace
+		}
+	}
+
+	if pvName == "" {
+		err := fmt.Errorf("CNS metadata for volume %s has missing pvName", instance.Spec.VolumeID)
+		log.Error(err)
+		return reconcile.Result{}, err
+	}
+
+	k8sclient, err := k8s.NewClient(ctx)
+	if err != nil {
+		log.Errorf("Failed to initialize K8S client when reconciling CnsUnregisterVolume "+
+			"instance: %s on namespace: %s. Error: %+v", instance.Name, instance.Namespace, err)
+		setInstanceError(ctx, r, instance, "Failed to init K8S client for volume unregistration")
+		return reconcile.Result{RequeueAfter: timeout}, nil
+	}
+
+	//Change PV ReclaimPolicy to retain so that underlying FCD doesn't get deleted when deleting PV,PVC
+	pv, err := k8sclient.CoreV1().PersistentVolumes().Get(ctx, pvName, metav1.GetOptions{})
+	if err != nil {
+		if !apierrors.IsNotFound(err) {
+			log.Errorf("Unable to get PV %q", pvName)
+			return reconcile.Result{}, err
+		}
+	}
+
+	if pv.Spec.PersistentVolumeReclaimPolicy != v1.PersistentVolumeReclaimRetain {
+		pv.Spec.PersistentVolumeReclaimPolicy = v1.PersistentVolumeReclaimRetain
+		retryErr := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+			_, updateErr := k8sclient.CoreV1().PersistentVolumes().Update(context.TODO(), pv, metav1.UpdateOptions{})
+			return updateErr
+		})
+		if retryErr != nil {
+			log.Errorf("Unable to update ReclaimPolicy on PV %q", pvName)
+			return reconcile.Result{}, err
+		}
+		log.Infof("Updated ReclaimPolicy on PV %q to %q", pvName, v1.PersistentVolumeReclaimRetain)
+	}
+
+	// Delete PVC.
+	if pvcName != "" && namespace != "" {
+		err = k8sclient.CoreV1().PersistentVolumeClaims(namespace).Delete(ctx,
+			pvcName, *metav1.NewDeleteOptions(0))
+		if err != nil {
+			if !apierrors.IsNotFound(err) {
+				log.Errorf("Failed to delete PVC %q in namespace %q with error - %s",
+					pvcName, namespace, err.Error())
+				return reconcile.Result{}, err
```

Comment:
Do you assume the CNS entry is deleted if the PVC is not found?

Reply:
We're setting the PV reclaim policy to `Retain` before deleting the PVC. So if only PVC gets deleted and some error occurs during PV/CNS volume deletion, the PV and CNS volume will still be there in the next reconcile loop.